### PR TITLE
Merkle tree changes

### PR DESCRIPTION
**Pull request by muXxer**
# Description of change

This PR fixes an issue in the MerkleTree.ReadFrom function and adds some additional progress callbacks to the MerkleCreateOptions + some code cleanups.

## Type of change

- Bug fix (a non-breaking change which fixes an issue)
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Implemented the new merkle lib in HORNET.

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: merkle/merkle_persistency.go [hunk 1]
```diff
@@ -11,8 +11,8 @@ import (
 )
 
 var (
-	errPersInvDepth  = errors.New("error parsing Merkle Tree: invalid depth")
-	errPersInvLayers = errors.New("error parsing Merkle Tree: invalid Layers length")
+	errPersInvDepth  = errors.New("error parsing merkle tree: invalid depth")
+	errPersInvLayers = errors.New("error parsing merkle tree: invalid layers length")
```
Comment (luca-moser) [inline merkle/merkle_persistency.go:15]: Please remove the lowercasing of Merkle.

File: merkle/merkle_persistency.go [hunk 2]
```diff
@@ -75,7 +75,7 @@ func (mt *MerkleTree) ReadFrom(buf *bufio.Reader) (n int64, err error) {
 	}
 	bytesRead += int64(binary.Size(&lengthLayers))
 
-	if lengthLayers != depth {
+	if lengthLayers != depth+1 {
```
Comment (luca-moser) [inline merkle/merkle_persistency.go:78]: explain why +1 with a comment

File: merkle/merkle_tree.go [hunk 3]
```diff
@@ -45,27 +47,45 @@ type MerkleTreeLayer struct {
 
 // MerkleCreateOptions is used to pass optional creation options to CreateMerkleTree.
 type MerkleCreateOptions struct {
-	// A callback that will be called after each leaf generation, with the corresponding index.
-	ProgressCallback func(uint32)
+	// CalculateAddressesStartCallback will be called at start of leaf generation, with the total count of the leafes.
```
Comment (luca-moser) [inline merkle/merkle_tree.go:50]:
```suggestion
	// CalculateAddressesStartCallback will be called at the start of leaf generation, with the total count of the leaves.
```

File: merkle/merkle_tree.go [hunk 4]
```diff
@@ -45,27 +47,45 @@ type MerkleTreeLayer struct {
...
+	CalculateAddressesStartCallback func(count uint32)
+	// CalculateAddressesCallback will be called after each leaf generation, with the corresponding index.
+	CalculateAddressesCallback func(index uint32)
+	// CalculateAddressesFinishedCallback will be called after leaf generation is finished, with the total count of the leafes.
```
Comment (luca-moser) [inline merkle/merkle_tree.go:54]:
```suggestion
	// CalculateAddressesFinishedCallback will be called after leaf generation is finished, with the total count of the leaves.
```

File: tools/merkle_calculator/merkle_tree_calculator.go [hunk 5]
```diff
@@ -13,7 +13,7 @@ import (
 func main() {
 
 	var (
-		depth         = flag.Int("depth", 0, "Depth of the Merkle tree to create")
+		depth         = flag.Int("depth", 0, "Depth of the merkle tree to create")
```
Comment (luca-moser) [inline tools/merkle_calculator/merkle_tree_calculator.go:16]: Remove this change.

File: tools/merkle_calculator/merkle_tree_calculator.go [hunk 6]
```diff
@@ -61,20 +63,34 @@ func main() {
...
+			CalculateAddressesStartCallback:    calculateAddressesStartCallback,
+			CalculateAddressesCallback:         calculateAddressesCallback,
+			CalculateAddressesFinishedCallback: calculateAddressesFinishedCallback,
+			CalculateLayersCallback:            calculateLayersCallback,
+			Parallelism:                        *parallelism,
+		})
 
 	if err != nil {
-		log.Panicf("Error creating Merkle tree: %v", err)
+		log.Panicf("Error creating merkle tree: %v", err)
```
Comment (luca-moser) [inline tools/merkle_calculator/merkle_tree_calculator.go:84]: remove change

File: tools/merkle_calculator/merkle_tree_calculator.go [hunk 7]
```diff
@@ -61,20 +63,34 @@ func main() {
...
 		return
 	}
 
 	if err := merkle.StoreMerkleTreeFile(*outputPath, mt); err != nil {
-		log.Panicf("Error persisting Merkle tree: %v", err)
+		log.Panicf("Error persisting merkle tree: %v", err)
```
Comment (luca-moser) [inline tools/merkle_calculator/merkle_tree_calculator.go:89]: remove change

File: tools/merkle_calculator/merkle_tree_calculator.go [hunk 8]
```diff
@@ -61,20 +63,34 @@ func main() {
...
 		return
 	}
 
-	log.Printf("Merkle tree root: %v\n", mt.Root)
-
-	log.Printf("Took %v seconds.\n", time.Since(ts).Truncate(time.Second))
+	log.Printf("merkle tree root: %v\n", mt.Root)
```
Comment (luca-moser) [inline tools/merkle_calculator/merkle_tree_calculator.go:93]: Merkle

File: tools/merkle_calculator/merkle_tree_calculator.go [hunk 9]
```diff
@@ -61,20 +63,34 @@ func main() {
...
 
+	log.Printf("successfully created merkle tree (took %v).\n", time.Since(ts).Truncate(time.Second))
```
Comment (luca-moser) [inline tools/merkle_calculator/merkle_tree_calculator.go:95]: Merkle

**Comment by Wollac** [comment]
@muXxer while you are at it, could you maybe also change the persistency to `WriteTo(io.Writer) (n int64, err error)` and `ReadFrom(io.Reader) (n int64, err error)`?
There is no need to force this to be an `*bufio.Reader` and we were aiming to implement the `ReaderFrom` `WriterTo` interface.
And this PR seems to be a good opportunity :sweat_smile:

**Pull request merged**